compiler/noirc_evaluator/src/ssa.rs

```diff
@@ -75,6 +75,31 @@ pub(crate) fn optimize_into_acir(
     time("SSA to ACIR", print_timings, || ssa.into_acir(&brillig, abi_distinctness))
 }
 
+/// Loop unrolling can return errors, since ACIR functions need to be fully unrolled.
+/// This meta-pass will keep trying to unroll loops and simplifying the SSA until no more errors are found.
+fn unroll_all_acir_loops(mut ssa: Ssa) -> Result<Ssa, RuntimeError> {
+    // Try to unroll loops first:
+    let mut unroll_errors;
+    (ssa, unroll_errors) = ssa.try_to_unroll_loops();
+
+    // Keep unrolling until no more errors are found
+    while !unroll_errors.is_empty() {
+        let prev_unroll_err_count = unroll_errors.len();
+
+        // Simplify the SSA before retrying
+        ssa = ssa.simplify_cfg();
```

Comment:
The linked issue being resolved just does a simple for loop. Without `flatten_cfg` I think we would fail if there are any if statements based off of witness conditions. If it is simple, perhaps we want to flatten here as well. Otherwise as this provides a quick fix for the linked issue I think we can keep this for now and handle that in a followup.

Reply:
Flattening is a global pass which is also incompatible if there are loops still in the program. If there are any if statements based off witness conditions the resulting index from them would still be unknown afterward anyway since they'd be dependent on the condition still.

Reply:
Yes, this would only fix it for loops whose bounds don't depend on ifs branches. Unfortunately, flatten_cfg fails if there are other sources of control flow than if branches (the remaning loops) so it cannot be used here in its current form at least

Reply:
> the resulting index from them would still be unknown afterward anyway since they'd be dependent on the condition still.

Ah yes. Ignore my comment.

Reply:
Oh wait! I think just mem2reg and simplify_cfg might be enough to deal with this

Reply:
If I just do the mem2reg before the simplification of the cfg, the simplification makes it so ifs with now-known branches are removed :)

Reply:
And probably a mem2reg after the simplification is not needed because mem2reg handles non-loop cfg well I think @jfecher

Reply:
Ah yeah that makes sense for compile-time ifs though. I was talking about witness conditionals which will still not be solvable.

Reply:
Yep but that makes sense, this is only for make unrolling work in code that has no runtime conditionals that changes the bounds